### PR TITLE
Layout system mutability refactor.

### PR DESCRIPTION
**Pull request by Britefury**
Refactored layout system so that element allocations are returned by value by allocation functions rather than by mutating existing `LAlloc` instances.

### CODE REVIEW

File: src/lib/lspace/layout/horizontal_layout.rs [hunk 1]
```diff
@@ -8,21 +8,27 @@ pub fn requisition_x(child_x_reqs: &[&LReq], x_spacing: f64) -> LReq {
...
+                    child_reqs: &[&LReq], x_spacing: f64) -> Vec<LAlloc> {
+    return LAlloc::alloc_linear(child_reqs, box_x_req, box_x_alloc.pos_in_parent(),
+                                box_x_alloc.alloc_size(), box_x_alloc.ref_point(), x_spacing, None);
 }
 
 pub fn requisition_y(child_y_reqs: &[&LReq]) -> LReq {
     return LReq::perpendicular_acc(child_y_reqs);
 }
 
+//pub fn alloc_y(box_y_req: &LReq, box_y_alloc: &LAlloc,
```
Comment (ltratt) [inline src/lib/lspace/layout/horizontal_layout.rs]: Remove commented code.

File: src/lib/lspace/layout/horizontal_layout.rs [hunk 2]
```diff
@@ -129,9 +119,7 @@ mod tests {
             children.push(LBox::new(LReq::new(size_x, flex_x), LReq::new(size_y, flex_y)))
         }
 
-        let mut child_refs : Vec<&mut LBox> = children.iter_mut().collect();
-        let (child_x_req_refs, mut child_x_alloc_and_req_refs, child_y_req_refs, mut child_y_alloc_and_req_refs) =
-                LBox::reqs_and_mut_allocs(&mut child_refs);
+//        let mut child_refs : Vec<&mut LBox> = children.iter_mut().collect();
```
Comment (ltratt) [inline src/lib/lspace/layout/horizontal_layout.rs]: Remove commented code.

File: src/lib/lspace/layout/lbox.rs [hunk 3]
```diff
@@ -23,39 +23,27 @@ impl LBox {
...
 
-    pub fn x_reqs_and_mut_allocs<'a>(boxes: &'a mut [&'a mut LBox]) -> Vec<(&'a LReq, &'a mut LAlloc)> {
-        return boxes.iter_mut().map(|b| (&b.x_req, &mut b.x_alloc)).collect();
-    }
-
-    pub fn y_reqs_and_mut_allocs<'a>(boxes: &'a mut [&'a mut LBox]) -> Vec<(&'a LReq, &'a mut LAlloc)> {
-        return boxes.iter_mut().map(|b| (&b.y_req, &mut b.y_alloc)).collect();
+    pub fn update_x_allocs(boxes: &mut [LBox], x_allocs: &[LAlloc]) {
+        let n = boxes.len();
+        assert!(n == x_allocs.len());
```
Comment (ltratt) [inline src/lib/lspace/layout/lbox.rs]: Should be `assert_eq` (gives better error messages).

File: src/lib/lspace/layout/lbox.rs [hunk 4]
```diff
@@ -23,39 +23,27 @@ impl LBox {
...
-        let mut y_reqs : Vec<&LReq> = Vec::with_capacity(n);
-        let mut x_reqs_and_allocs : Vec<(&LReq, &mut LAlloc)> = Vec::with_capacity(n);
-        let mut y_reqs_and_allocs : Vec<(&LReq, &mut LAlloc)> = Vec::with_capacity(n);
-
-        for (mut x_alloc, x_req, mut y_alloc, y_req) in everything {
-            x_reqs.push(x_req);
-            y_reqs.push(y_req);
-            x_reqs_and_allocs.push((x_req, x_alloc));
-            y_reqs_and_allocs.push((y_req, y_alloc));
+        assert!(n == y_allocs.len());
```
Comment (ltratt) [inline src/lib/lspace/layout/lbox.rs]: `assert_eq`

File: src/lib/lspace/layout/vertical_layout.rs [hunk 5]
```diff
@@ -133,9 +114,7 @@ mod tests {
             children.push(LBox::new(LReq::new(size_x, flex_x), LReq::new(size_y, flex_y)))
         }
 
-        let mut child_refs : Vec<&mut LBox> = children.iter_mut().collect();
-        let (child_x_req_refs, mut child_x_alloc_and_req_refs, child_y_req_refs, mut child_y_alloc_and_req_refs) =
-                LBox::reqs_and_mut_allocs(&mut child_refs);
+//        let mut child_refs : Vec<&mut LBox> = children.iter_mut().collect();
```
Comment (ltratt) [inline src/lib/lspace/layout/vertical_layout.rs]: Remove commented code.

**Comment by Britefury** [comment]
Ready for merge.

**Pull request merged**